platform-sdk/docs/proposals/TSS-Block-Signing/tss-block-signing.md

```diff
@@ -0,0 +1,309 @@
+not user generated. They should be gossiped with the highest priority.
+
+It may be possible that a node is aggregating share signature transactions from multiple ledger signature requests.
+In addition to the `share_signature`, the `roster_hash` and `share_index` are needed to identify the share public
+key to use for validation. The `message_hash` is needed to identify the message that was signed. A ledger signature
+can be produced by a threshold number of `share_signatures` for the same message. The threshold value for
+aggregation is determined by the roster indicated by the `roster_hash`.
+
+```protobuf
+message TssShareSignatureTransaction {
+  /**
+   * The hash of the roster containing the node whose share produced the share signature.
+   */
+  bytes roster_hash = 1;
+
+  /**
+   * The index of the share that produced this share signature.
+   */
+  uint64 share_index = 2;
+
+  /**
+   * The hash of the message that was signed by the share.
+   */
+  bytes message_hash = 3;
+
+  /**
+   * The share signature produced by the share.
+   */
+  bytes share_signature = 4;
+}
+```
+
+### Architecture and/or Components
+
+In addition to the extension of the `TssBaseService` interface, the service's `TransactionHandler` is extended to
+handle `TssShareSignatureTransaction` in the `pre-handle` phase.
+
+### Core Behaviors
+
+Thread Execution:
+
+* All code related to requesting a ledger signature,
+  generating share signature transactions, and handling share signature transactions do not modify the state and
+  should not execute on the `transaction handling thread`.
+* As soon as the ledger signature for a message can be constructed, all registered consumers are notified with the
+  message hash and the ledger signature. If there are multiple consumers registered, whether they are given the
+  signature serially or in parallel makes no difference as the data is immutable and consumption is expected to be
+  idempotent.
+
+Ledger Signature Generation:
+
+* The storage of `TssShareSignatureTransaction` is only in memory and expires after a configured amount of time.
+* Should a node be reset in the middle of the collection of share signatures, most signatures will be recoverable
+  through PCES replay.
+* If through restart or reconnect, a node is not able to collect enough share signatures to produce a ledger signature,
+  the accumulated share signatures will be thrown away after their holding time expires. No notification is generated
+  about the failure to produce a ledger signature.
+  * This infrequent failure to produce a ledger signature is expected to be rare and tolerated by ledger signature
+    consumers.
+  * Failures will be logged at an `INFO` level and metrics will be kept to track their occurrence.
+
+### Configuration
+
+New Configuration:
+
+* `tss.signatureLivenessPeriod` - The amount of time a share signature is held in memory before being discarded.
```

Comment:
I'm not sure we need this configuration value. Shouldn't it be enough to keep share signatures until either

1. an aggregate signature is produced for the block
2. the block is emitted without a ledger signature

and immediately delete once either of these conditions are met? This guarantees we will have share signatures as long as they are required, and no longer.

Reply:
This assumes that block signing will be the only use case for ledger signatures.  I would rather not create a call back mechanism where block signing has to notify the `TssBaseService` that a block was sent without a signature.

Reply:
It wouldn't need to be a call explicitly saying "block sent without a signature". It could just be an additional method saying `nevermindAboutThisSigningRequest`

There a couple reasons I think this would be a preferable solution:

1. it's simpler to implement than a time based retention mechanism
2. it minimizes number of share signatures that must be kept
3. This alternate solution is equally compatible with future ledger signature use cases: simply keep share signatures either until the aggregate is produced, or `nevermindAboutThisSigningRequest` is called

Reply:
It wouldn't need to be a call explicitly saying "block sent without a signature". It could just be an additional method saying `nevermindAboutThisSigningRequest`

There a couple reasons I think this would be a preferable solution:

1. it minimizes number of share signatures that must be kept
2. going this route would simplify the initial implementation, since block signatures wouldn't need to have a time-based cleanup. if a time-based cleanup is ever required for a future use-case, it could be implemented at that time

Reply:
Suppose the call back never occurs and the signature is not created because a threshold of share signatures were not aggregated.  That's a memory leak at that point without a timeout.   I don't see how to safely avoid not needing a timeout.   We could infer a timeout instead of having a configuration value based on later signature requests completing and their share signatures being garbage collected.